File: 0146-lru-cache/0146-lru-cache.java
public class Node {
  protected int key;
  protected int val;
  protected Node prev;
  protected Node next;
  
  public Node(int key, int val) {
    this.key = key;
    this.val = val;
    this.prev = null;
    this.next = null;
  }
}

class LRUCache {
  protected int capacity;
  protected Map<Integer, Node> cache;
  protected Node left;
  protected Node right;
  
  public LRUCache(int capacity) {
    this.capacity = capacity;
    this.cache = new HashMap<Integer, Node>();
    this.left = new Node(0, 0);
    this.right = new Node(0, 0);
    this.left.next = this.right;
    this.right.prev = this.left;
  }

  public void remove(Node node) {
    Node prev = node.prev;
    Node next = node.next;
    prev.next = next;
    next.prev = prev;
  }
  
  public void insert(Node node) {
    Node prev = this.right.prev;
    Node next = this.right;
    prev.next = node;
    next.prev = node;
    node.next = next;
    node.prev = prev;
  }

  public int get(int key) {
    if (this.cache.containsKey(key)) {
      remove(this.cache.get(key));
      insert(this.cache.get(key));
      return this.cache.get(key).val;
    }
    return -1;
  }

  public void put(int key, int value) {
    if (this.cache.containsKey(key)) remove(this.cache.get(key));
    this.cache.put(key, new Node(key, value));
    insert(this.cache.get(key));
    
    if (this.cache.size() > this.capacity) {
      Node lru = this.left.next;
      remove(lru);
      this.cache.remove(lru.key);
    }
  }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.rb
class Node
  attr_accessor :key, :val, :prev, :next
  
=begin
  :type key: Integer
  :type val: Integer
=end
  def initialize(key, val)
    @key = key
    @val = val
    @prev = @next = nil
  end
  
end

class LRUCache

=begin
  :type capacity: Integer
=end
  def initialize(capacity)
    @capacity = capacity
    @cache = Hash.new
    @left, @right = Node.new(0, 0), Node.new(0, 0)
    @left.next = @right
    @right.prev = @left
  end

  
=begin
  :type node: Node
  :rtype: Void
=end
  def remove(node)
    prev, nxt = node.prev, node.next
    prev.next, nxt.prev = nxt, prev
  end

  
=begin
  :type node: Node
  :rtype: Void
=end
  def insert(node)
    prev, nxt = @right.prev, @right
    prev.next = nxt.prev = node
    node.next, node.prev = nxt, prev
  end

  
=begin
  :type key: Integer
  :rtype: Integer
=end
  def get(key)
    if @cache.include?(key)
      remove(@cache[key])
      insert(@cache[key])
      return @cache[key].val
    end
    -1
  end


=begin
  :type key: Integer
  :type value: Integer
  :rtype: Void
=end
  def put(key, value)
    remove(@cache[key]) if @cache.include?(key)
    @cache[key] = Node.new(key, value)
    insert(@cache[key])
    
    if @cache.length > @capacity
      lru = @left.next
      remove(lru)
      @cache.delete(lru.key)
    end
  end


end

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache.new(capacity)
# param_1 = obj.get(key)
# obj.put(key, value)